internal/controller/nemoservice_controller.go

```diff
@@ -0,0 +1,586 @@
+// move the current state of the cluster closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.18.2/pkg/reconcile
+func (r *NemoServiceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+	defer r.refreshMetrics(ctx)
+
+	// Fetch the NemoService instance
+	nemoService := &appsv1alpha1.NemoService{}
+	if err := r.Get(ctx, req.NamespacedName, nemoService); err != nil {
+		if client.IgnoreNotFound(err) != nil {
+			logger.Error(err, "unable to fetch NemoService", "name", req.NamespacedName)
+		}
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	logger.Info("Reconciling", "NemoService", nemoService.Name)
+
+	// Check if the instance is marked for deletion
+	if nemoService.ObjectMeta.DeletionTimestamp.IsZero() {
+		// Add finalizer if not present
+		if !controllerutil.ContainsFinalizer(nemoService, NemoServiceFinalizer) {
+			controllerutil.AddFinalizer(nemoService, NemoServiceFinalizer)
+			if err := r.Update(ctx, nemoService); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	} else {
+		// The instance is being deleted
+		if controllerutil.ContainsFinalizer(nemoService, NemoServiceFinalizer) {
+			// Perform cleanup of resources
+			if err := r.cleanupNemoService(ctx, nemoService); err != nil {
+				return ctrl.Result{}, err
+			}
+
+			// Remove finalizer to allow for deletion
+			controllerutil.RemoveFinalizer(nemoService, NemoServiceFinalizer)
+			if err := r.Update(ctx, nemoService); err != nil {
+				return ctrl.Result{}, err
+			}
+			return ctrl.Result{}, nil
+		}
+	}
+
+	// Handle nemo services reconciliation
+	if result, err := r.reconcileNemoServices(ctx, nemoService); err != nil {
+		logger.Error(err, "error reconciling Nemo services", "name", nemoService.Name)
+		return result, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *NemoServiceReconciler) cleanupNemoService(ctx context.Context, nemoService *appsv1alpha1.NemoService) error {
+	// TODO: Add custom cleanup
+
+	// All owned objects are garbage collected
+	return nil
+}
+
+// Reconcile all enabled and disabled Nemo services
+func (r *NemoServiceReconciler) reconcileNemoServices(ctx context.Context, nemoService *appsv1alpha1.NemoService) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+	logger.Info("Reconciling Nemo services", "name", nemoService.Name)
+
+	enabledServices := make(map[string]bool)
+
+	// Define service reconcilers
+	serviceReconcilers := map[ServiceType]func(context.Context, *NemoServiceReconciler, *appsv1alpha1.NemoService) error{
+		Customizer: func(ctx context.Context, r *NemoServiceReconciler, ns *appsv1alpha1.NemoService) error {
+			spec := NemoServiceSpec[appsv1alpha1.NemoServiceCustomizerSpec]{
+				Enabled: ns.Spec.Customizer.Enabled,
+				Name:    ns.Spec.Customizer.Name,
+				Spec:    ns.Spec.Customizer,
+			}
+			enabledServices[spec.Name] = *spec.Enabled
```

Comment:
nit here and other references below:
Need to check for nil before dereferencing this ptr.
```suggestion
			enabledServices[spec.Name] = spec.Enabled != nil && *spec.Enabled
```